src/common/CryptoUtils.ts

```diff
@@ -0,0 +1,155 @@
+     * @param {number} keySize the size for the private key in bits, which must
+     * be between 2048 and 16384. Defaults to 4096.
+     * @param {number | Date} validity the validity length for the certificate.
+     * It can either be a number (representing the number of days from today) or
+     * it can be a specific date (which has to be at least 24 hours from now).
+     * Defaults to 365 days. The maximum length cannot be more than 825 days in the
+     * future to meet Apple guidelines: https://support.apple.com/en-gb/103769
+     * @returns the generated certificate and key.
+     */
+    public static generateSelfSignedCert(
+        hostname: string = 'localhost',
+        keySize: number = 4096,
+        validity: number | Date = 365
+    ): PEMCertificate {
+        if (keySize < 2048 || keySize > 16_384) {
+            throw new Error(messages.getMessage('error:invalidKeySize'));
+        }
+
+        if (validity instanceof Date) {
+            const millisecondsInOneDay = 24 * 60 * 60 * 1000;
+            const diff = validity.getTime() - new Date().getTime();
+            if (diff < millisecondsInOneDay || diff > 825 * millisecondsInOneDay) {
+                throw new Error(messages.getMessage('error:invalidValidityDate'));
+            }
+        } else if (validity < 1 || validity > 825) {
+            throw new Error(messages.getMessage('error:invalidValidityNumber'));
+        }
+
+        const keys = forge.pki.rsa.generateKeyPair(keySize);
+        const cert = forge.pki.createCertificate();
+
+        const startDate = new Date();
+        let endDate = new Date();
+        if (validity instanceof Date) {
+            endDate = validity;
+        } else {
+            endDate.setDate(startDate.getDate() + validity);
+        }
+
+        cert.publicKey = keys.publicKey;
+        cert.serialNumber = '01';
+        cert.validity.notBefore = startDate;
+        cert.validity.notAfter = endDate;
+
+        const attrs = [
+            { name: 'commonName', value: hostname },
+            { name: 'countryName', value: 'US' },
+            { shortName: 'ST', value: 'California' },
+            { name: 'localityName', value: 'San Francisco' },
+            { name: 'organizationName', value: 'Salesforce Inc.' },
+            { shortName: 'OU', value: 'LocalDevPreview' }
+        ];
+
+        cert.setSubject(attrs);
+        cert.setIssuer(attrs);
+
+        // Add the subjectAltName and ExtendedKeyUsage extensions, which are required by Apple
+        cert.setExtensions([
+            {
+                name: 'basicConstraints',
+                cA: true
+            },
+            {
+                name: 'keyUsage',
+                keyCertSign: true,
+                digitalSignature: true,
+                nonRepudiation: true,
+                keyEncipherment: true,
+                dataEncipherment: true
+            },
+            {
+                name: 'extKeyUsage',
```

Comment:
I don't think this section is necessary.

Reply:
Oh maybe for macOS? I know I get similar prompts when I import keys into the Keychain on my computer.

Reply:
I put most of these here b/c I saw it in the examples provided by `node-forge` and as you pointed out we probably can remove most of these altogether.

For the cert to be valid on iOS it needs to meet [these requirements](https://support.apple.com/en-gb/103769).

Reply:
> For the cert to be valid on iOS it needs to meet [these requirements](https://support.apple.com/en-gb/103769).

Ah, didn't know about the newer EKU requirements. They don't seem to impact my usage, without them. But I'm fine with going with those guidelines, since they're out there and all. We could probably just get away with the `serverAuth` entry, out of the ones we have at the moment, per their requirements.